---
title: "integration"
format: html
editor: visual
---

```{r}
library(DESeq2)
library(Seurat)
library("textTinyR")
library(tidyverse)
library(ggrepel)
library(BPCells)
library(tidyseurat)
library(SeuratWrappers)
library(extrafont)
#font_import()
loadfonts(device = "win")
```

```{r}
path_data_file <- "C:\\Users\\danne\\raw_data\\cubi_cusco\\"
path_output_file <- "output\\"

dataset_name <- "hub_01"

library(future)
plan("multicore", workers = 8)
options(future.globals.maxSize = 1300 * 1024^2)
```

```{r}
obj <- read_rds( file = paste0("intermediate_data/",dataset_name,"_filtered_seurat_obj_processed.rds" ))

obj <- read_rds(paste0("intermediate_data/", dataset_name, "_integrated_", ".rds"))
                
DefaultAssay(obj) <- "RNA"
obj[["RNA"]] <- split(obj[["RNA"]], f = obj$cell_line_cellranger)
obj
```

```{r}
#perfrom standard data processing on layers

obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)
obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
obj <- FindClusters(obj, resolution = 2, cluster.name = "unintegrated_clusters")
```

```{r}
if ( FALSE # not calculated because of exessive cumputing times
  #!"integrated.cca" %in% names(obj@reductions)
    ) {
  

obj <- IntegrateLayers(
  object = obj, method = CCAIntegration,
  orig.reduction = "pca", new.reduction = "integrated.cca",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "integrated.cca", dims = 1:30)
obj <- FindClusters(obj, resolution = 2, cluster.name = "cca_clusters")
obj <- RunUMAP(obj, reduction = "integrated.cca", dims = 1:30, reduction.name = "umap.cca")

write_rds(obj, "../intermediate_data/schulte_schrepping_10x_berlin.rds")

  
}


if ( FALSE # not calculated because of exessive cumputing times
  #!"integrated.rpca" %in% names(obj@reductions)
    ) {
  obj <- IntegrateLayers(
  object = obj, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = "integrated.rpca",
  verbose = FALSE,
)
write_rds(obj, "../intermediate_data/schulte_schrepping_10x_berlin.rds")
  
}


if ( !"umap.harmony" %in% names(obj@reductions)
    ) {
  obj <- IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)

obj <- FindNeighbors(obj, reduction = "harmony", dims = 1:30)
obj <- FindClusters(obj,resolution = 0.2, cluster.name = "harmony_clusters")

obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")
Idents(obj) <- "cluster_clean"
write_rds(obj, "../intermediate_data/schulte_schrepping_10x_berlin.rds")
}
```

```{r}
if ( !"umap.mnn" %in% names(obj@reductions)
    ) {
  obj <- IntegrateLayers(
  object = obj, method = FastMNNIntegration,
  new.reduction = "integrated.mnn",
  verbose = FALSE, assay = "RNA"
  )
  obj <- FindNeighbors(obj, reduction = "integrated.mnn", dims = 1:30)
  obj <- FindClusters(obj,resolution = 0.2, cluster.name = "mnn_clusters")
  obj <- FindClusters(obj,resolution = 0.1, cluster.name = "harmony_clusters_res0.1")
  obj <- FindClusters(obj,resolution = 0.4, cluster.name = "harmony_clusters_res0.4")
  obj <- RunUMAP(obj, reduction = "integrated.mnn", dims = 1:30, reduction.name = "umap.mnn")
  obj <- RunPCA(obj, reduction = "integrated.mnn", dims = 1:30, reduction.name = "pca_mnn")
  
  write_rds(obj, paste0("intermediate_data/", dataset_name, "_integrated_", ".rds"))
}



# obj <- IntegrateLayers(
#   object = obj, method = SeuratWrappers::scVIIntegration,
#   new.reduction = "integrated.scvi",
#   #conda_env = "../miniconda3/envs/scvi-env", 
#   verbose = FALSE
# )
```

```{r}
grouping <- "harmony_clusters_res0.1"
red <- "umap.mnn"
obj |> DimPlot(reduction = red, label = T, group.by = grouping )+ coord_fixed()
ggsave(paste0(path_output_file, "hub_01_dimplot_",red, "_", grouping, ".png"))


grouping <- "harmony_clusters_res0.4"

obj |> DimPlot(reduction = red, label = T, group.by = grouping )+ coord_fixed()
ggsave(paste0(path_output_file, "hub_01_dimplot_",red, "_", grouping, ".png"))

obj |> DimPlot(reduction = "umap.mnn", label = T, group.by = "harmony_clusters_res0.4" )


grouping <- "cell_line_cellranger"
obj |> DimPlot(reduction = red, group.by = grouping )+ coord_fixed()
ggsave(paste0(path_output_file, "hub_01_dimplot_",red, "_", grouping, ".png"))


grouping <- "Phase"
obj<- CellCycleScoring(obj, s.features = cc.genes$s.genes,  g2m.features = cc.genes$g2m.genes)
obj |> DimPlot(reduction = "umap.mnn", group.by = "Phase")+ coord_fixed()
ggsave(paste0(path_output_file, "hub_01_dimplot_",red, "_", grouping, ".png"))
```

```{r}
#make plots for probematic genes for deconv
```

```{r}
obj |> ggplot(aes(cell_line_cellranger, fill=harmony_clusters_res0.4)) + geom_bar()
ggsave(paste0(path_output_file, "hub_01_bar_absolute_","harmony_clusters_res0.4_", ".png"))

obj |>
  select(cell_line_cellranger,harmony_clusters_res0.4) |> 
  group_by(cell_line_cellranger,harmony_clusters_res0.4) |> 
  summarise(n=n()) |> 
  group_by(cell_line_cellranger) |> 
  mutate(percentage = n / sum(n) * 100) |> 
  ggplot(aes(x = cell_line_cellranger, y = percentage, fill = harmony_clusters_res0.4)) +
  geom_bar(stat = "identity") +
  labs(y = "Percentage (%)", x = "Category", title = "Cluster composition of samples") +
  theme_minimal()

ggsave(paste0(path_output_file, "hub_01_bar_relative_","harmony_clusters_res0.4_", ".png"))


```

```{r}
obj <- JoinLayers(obj)
markers_mnn <- FindAllMarkers(obj)
#markers_mnn_roc <- FindAllMarkers(obj, test.use = "roc", logfc.threshold = 2)

top_m <- markers_mnn  |> group_by(cluster) |> slice_min(p_val_adj, n=20) |>  slice_max(avg_log2FC, n=5)
DotPlot(object = obj
        ,features = pull(top_m, gene) |> unique(), group.by = "harmony_clusters_res0.4")+coord_flip()

ggsave(paste0(path_output_file, "hub_01_dotPlot_markers_","harmony_clusters_res0.4_", ".png"), width = 8, height = 15)

```

```{r}
FeaturePlot(obj, features = c("EPCAM" ,"TFF1", "LYZ","VIM", "SOX3","SVOP","EYA2", "TOP2A", "RPL9"),
            reduction = "umap.mnn") 

ggsave(paste0(path_output_file, "hub_01_feature_some_markers","harmony_clusters_res0.4_", ".png"), width = 10, height = 8)

```

## NCBI Gene Summary for TFF3 Gene

Members of the trefoil family are characterized by having at least one copy of the trefoil motif, a 40-amino acid domain that contains three conserved disulfides. They are stable secretory proteins expressed in gastrointestinal mucosa. Their functions are not defined, but they may protect the mucosa from insults, stabilize the mucus layer and affect healing of the epithelium. This gene is expressed in goblet cells of the intestines and colon. This gene and two other related trefoil family member genes are found in a cluster on chromosome 21. \[provided by RefSeq, Jul 2008\]

CIViC Summary for TFF3 Gene GeneCards Summary for TFF3 Gene TFF3 (Trefoil Factor 3) is a Protein Coding gene. Diseases associated with TFF3 include Cardiomyopathy, Familial Hypertrophic, 15 and Cervical Adenofibroma. Among its related pathways are ESR-mediated signaling and Pancreatic cancer subtypes. An important paralog of this gene is TFF1.

UniProtKB/Swiss-Prot Summary for TFF3 Gene Involved in the maintenance and repair of the intestinal mucosa. Promotes the mobility of epithelial cells in healing processes (motogen). ( TFF3_HUMAN,Q07654 )

## NCBI Gene Summary for LYZ Gene

This gene encodes human lysozyme, whose natural substrate is the bacterial cell wall peptidoglycan (cleaving the beta\[1-4\]glycosidic linkages between N-acetylmuramic acid and N-acetylglucosamine). Lysozyme is one of the antimicrobial agents found in human milk, and is also present in spleen, lung, kidney, white blood cells, plasma, saliva, and tears. The protein has antibacterial activity against a number of bacterial species. Missense mutations in this gene have been identified in heritable renal amyloidosis. \[provided by RefSeq, Oct 2014\]

GeneCards Summary for LYZ Gene LYZ (Lysozyme) is a Protein Coding gene. Diseases associated with LYZ include Amyloidosis, Hereditary Systemic 5 and Amyloidosis, Hereditary Systemic 2. Among its related pathways are Innate Immune System and Metabolism of proteins. Gene Ontology (GO) annotations related to this gene include identical protein binding and lysozyme activity. An important paralog of this gene is SPACA3.

UniProtKB/Swiss-Prot Summary for LYZ Gene Lysozymes have primarily a bacteriolytic function; those in tissues and body fluids are associated with the monocyte-macrophage system and enhance the activity of immunoagents. ( LYSC_HUMAN,P61626 )

## NCBI Gene Summary for VIM Gene

This gene encodes a type III intermediate filament protein. Intermediate filaments, along with microtubules and actin microfilaments, make up the cytoskeleton. The encoded protein is responsible for maintaining cell shape and integrity of the cytoplasm, and stabilizing cytoskeletal interactions. This protein is involved in neuritogenesis and cholesterol transport and functions as an organizer of a number of other critical proteins involved in cell attachment, migration, and signaling. Bacterial and viral pathogens have been shown to attach to this protein on the host cell surface. Mutations in this gene are associated with congenital cataracts in human patients. \[provided by RefSeq, Aug 2017\]

GeneCards Summary for VIM Gene VIM (Vimentin) is a Protein Coding gene. Diseases associated with VIM include Cataract 30, Multiple Types and Cataract 30. Among its related pathways are Selective autophagy and Signaling by Rho GTPases. Gene Ontology (GO) annotations related to this gene include identical protein binding and obsolete protein C-terminus binding. An important paralog of this gene is DES.

UniProtKB/Swiss-Prot Summary for VIM Gene Vimentins are class-III intermediate filaments found in various non-epithelial cells, especially mesenchymal cells. Vimentin is attached to the nucleus, endoplasmic reticulum, and mitochondria, either laterally or terminally. ( VIME_HUMAN,P08670 )

Involved with LARP6 in the stabilization of type I collagen mRNAs for CO1A1 and CO1A2. ( VIME_HUMAN,P08670 )

## NCBI Gene Summary for SOX3 Gene

This gene encodes a member of the SOX (SRY-related HMG-box) family of transcription factors involved in the regulation of embryonic development and in the determination of the cell fate. The encoded protein may act as a transcriptional regulator after forming a protein complex with other proteins. Mutations in this gene have been associated with X-linked cognitive disability with growth hormone deficiency. \[provided by RefSeq, Jul 2008\]

GeneCards Summary for SOX3 Gene SOX3 (SRY-Box Transcription Factor 3) is a Protein Coding gene. Diseases associated with SOX3 include Intellectual Developmental Disorder, X-Linked, With Panhypopituitarism and Panhypopituitarism, X-Linked. Among its related pathways are Signaling by WNT and ERK Signaling. Gene Ontology (GO) annotations related to this gene include DNA-binding transcription factor activity and transcription corepressor activity. An important paralog of this gene is SOX1.

UniProtKB/Swiss-Prot Summary for SOX3 Gene Transcription factor required during the formation of the hypothalamo-pituitary axis. May function as a switch in neuronal development. Keeps neural cells undifferentiated by counteracting the activity of proneural proteins and suppresses neuronal differentiation. Required also within the pharyngeal epithelia for craniofacial morphogenesis. Controls a genetic switch in male development. Is necessary for initiating male sex determination by directing the development of supporting cell precursors (pre-Sertoli cells) as Sertoli rather than granulosa cells (By similarity). ( SOX3_HUMAN,P41225 )

#3 NCBI Gene Summary for EYA2 Gene This gene encodes a member of the eyes absent (EYA) family of proteins. The encoded protein may be post-translationally modified and may play a role in eye development. A similar protein in mice can act as a transcriptional activator. Alternative splicing results in multiple transcript variants, but the full-length natures of all of these variants have not yet been determined. \[provided by RefSeq, Jul 2009\]

GeneCards Summary for EYA2 Gene EYA2 (EYA Transcriptional Coactivator And Phosphatase 2) is a Protein Coding gene. Diseases associated with EYA2 include Bruck Syndrome and Familial Adenomatous Polyposis 1. Among its related pathways are HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) and DNA Double Strand Break Response. Gene Ontology (GO) annotations related to this gene include magnesium ion binding and protein tyrosine phosphatase activity. An important paralog of this gene is EYA1.

UniProtKB/Swiss-Prot Summary for EYA2 Gene Functions both as protein phosphatase and as transcriptional coactivator for SIX1, and probably also for SIX2, SIX4 and SIX5 (PubMed:12500905, 23435380). Tyrosine phosphatase that dephosphorylates 'Tyr-142' of histone H2AX (H2AXY142ph) and promotes efficient DNA repair via the recruitment of DNA repair complexes containing MDC1. 'Tyr-142' phosphorylation of histone H2AX plays a central role in DNA repair and acts as a mark that distinguishes between apoptotic and repair responses to genotoxic stress (PubMed:19351884). Its function as histone phosphatase may contribute to its function in transcription regulation during organogenesis. Plays an important role in hypaxial muscle development together with SIX1 and DACH2; in this it is functionally redundant with EYA1 (PubMed:12500905). ( EYA2_HUMAN,O00167 )

## Summaries for STMN2 Gene

NCBI Gene Summary for STMN2 Gene This gene encodes a member of the stathmin family of phosphoproteins. Stathmin proteins function in microtubule dynamics and signal transduction. The encoded protein plays a regulatory role in neuronal growth and is also thought to be involved in osteogenesis. Reductions in the expression of this gene have been associated with Down's syndrome and Alzheimer's disease. Alternatively spliced transcript variants have been observed for this gene. A pseudogene of this gene is located on the long arm of chromosome 6. \[provided by RefSeq, Nov 2010\]

GeneCards Summary for STMN2 Gene STMN2 (Stathmin 2) is a Protein Coding gene. Diseases associated with STMN2 include Trichorhinophalangeal Syndrome, Type Ii and Malignant Syringoma. Among its related pathways are Signaling by Rho GTPases and RND3 GTPase cycle. Gene Ontology (GO) annotations related to this gene include calcium-dependent protein binding and tubulin binding. An important paralog of this gene is STMN3.

UniProtKB/Swiss-Prot Summary for STMN2 Gene Regulator of microtubule stability. When phosphorylated by MAPK8, stabilizes microtubules and consequently controls neurite length in cortical neurons. In the developing brain, negatively regulates the rate of exit from multipolar stage and retards radial migration from the ventricular zone (By similarity). ( STMN2_HUMAN,Q93045 )

```{r}

```
